mycpp/gc_dict.h

```diff
@@ -307,49 +283,92 @@ void Dict<K, V>::clear() {
   len_ = 0;
 }
 
-// Returns the position in the array.  Used by dict_contains(), index(),
-// get(), and set().
-//
-// For now this does a linear search.
 // TODO:
-// - hash functions, and linear probing.
-// - resizing based on load factor
-//   - which requires rehashing (re-insert all items)
 // - Special case to intern Str* when it's hashed?  How?
 //   - Should we have wrappers like:
 //   - V GetAndIntern<V>(D, &string_key)
 //   - SetAndIntern<V>(D, &string_key, value)
 //   This will enable duplicate copies of the string to be garbage collected
 template <typename K, typename V>
-int Dict<K, V>::position_of_key(K key) {
-  for (int i = 0; i < capacity_; ++i) {
-    int special = entry_->items_[i];  // NOT an index now
-    if (special == kDeletedEntry) {
-      continue;  // keep searching
+int Dict<K, V>::hash_and_probe(K key) const {
+  if (capacity_ == 0) {
+    return kNotFound;
+  }
+
+  // Hash the key onto a slot in the index. If the first slot is occupied, probe
+  // until an empty one is found.
+  unsigned h = hash_key(key);
+  int init_bucket = h % index_len_;
+
+  for (int i = 0; i < index_len_; ++i) {
+    // Start at init_bucket and wrap araound
+    int slot = (i + init_bucket) % index_len_;
```

Comment:
I read this modulus can be slow ... and we also kinda do it twice on lines 301 and 305

It depends if `index_len_` is a power of 2, and right now we're growing it by 3/2

I think this is OK now but it might come up later, when we tune the growth factors

Reply:
Yeah, it can be expensive, but it doesn't show up in the profile right now (I think we have a few other bottlenecks elsewhere). If it becomes a problem we can split this loop back out into two loops over `[init_bucket, index_len)` and `[0, init_bucket)` like before. I like how compact the single loop is as long as we can get away with it